belief updater that stacks the k previous observations

MaximeBouton opened the pull request.

zsunberg commented on src/beliefs/k_previous_observations.jl:
```diff
@@ -0,0 +1,21 @@
+    k::Int64
+end
+
+function initialize_belief{O}(bu::KMarkovUpdater, obs::O, b=nothing)
```
We should get rid of the `b=nothing`. That's not part of POMDPs.jl anymore and I think it will only confuse people.

MaximeBouton commented:
Thanks for the feedback, I ll make the changes and add some documentation as well

zsunberg commented:
Actually, on second thought, we shouldn't use a circular buffer because that would overwrite old beliefs. The cost of copying should not actually be too big.

MaximeBouton commented:
Here is a version using `CircularBuffer`, @zsunberg  do you think this is more elegant?
```julia
struct KMarkovUpdater <: Updater
    k::Int
end

function initialize_belief{O}(bu::KMarkovUpdater, obs::O)
    b0 = CircularBuffer{O}(bu.k)
    for i=1:bu.k
        push!(b0, obs)
    end
    @assert isfull(b0)
    return b0
end

function update{O}(bu::KMarkovUpdater, old_b::CircularBuffer{O}, action, obs::O)
    new_b = CircularBuffer{O}(bu.k)
    append!(new_b, old_b)
    push!(new_b, obs)
    @assert isfull(new_b)
    @assert capacity(new_b) == bu.k
    return new_b
end
```

zsunberg commented:
@MaximeBouton Oh shoot, sorry! I don't think it makes sense to use the circular buffer because we don't want to change previous beliefs, so we have to copy anyways (as you are doing). I think we should just go back to the normal Vector.

zsunberg commented:
@MaximeBouton Can you add a test that runs `simulate` with a `HistoryRecorder`? It seems like we should make something that works with a `HistoryRecorder`, but there might be a few obstacles to overcome. Let me know if there are any problems.

MaximeBouton commented:
Yes, I tried adding the buffer in the updater but since we require `b_old` to be unchanged and `update(up, b_old, a, o)` to always return the same thing it was not really possible to maintain a buffer.

MaximeBouton commented:
It works with `simulate` and `HistoryRecorder` but the usage is a bit weird: 
```julia 
pomdp = BabyPOMDP()
solver = RandomSolver(rng=rng)
policy = solve(solver, pomdp)

s0 = initial_state(pomdp, rng)
hr = HistoryRecorder(rng=rng, initial_state = s0, max_steps=100)
o0 = generate_o(pomdp, s0, rng)
hist = simulate(hr, pomdp, policy, up, o0)
```

First, one must make sure that the initial observation is consistent with the initial state. 
Since `simulate` is calling `initialize_belief(bu::Updater, init_state_distribution::Any)` , the user must pass the observation rather than an initial state distribution which is a bit counter-intuitive.

zsunberg commented:
Yeah, it is a little bit nonstandard, but I think it is ok to have to pass in the observation as the "belief". I think the most important thing is to output a useful error message if it's not done right (See also JuliaPOMDP/POMDPToolbox.jl#56 about the initial state - that might make things less confusing).

The other option is to seed the updater with an initial observation, but that seems more confusing.

We should also allow passing in a vector of `k` previous observations if people want to initialize it like that.

MaximeBouton commented:
I added tests with a history recorder as well as a method to initialize the belief using a vector. 

I am not sure how to implement the error msg, what should it check? 
I added an assertion in `initialize_belief{O}(up, obs_vec::Vector{O})` to check that the size matches the parameter in the updater.

zsunberg commented on src/POMDPToolbox.jl:
```diff
@@ -11,11 +11,13 @@ import POMDPs: action, value, solve
 import POMDPs: actions, action_index, state_index, obs_index, iterator, sampletype, states, n_actions, n_states, observations, n_observations, discount, isterminal
 import POMDPs: generate_sr, initial_state
 import Base: rand, rand!, mean, ==
+import DataStructures: CircularBuffer, isfull, capacity, push!, append!
```
We don't need this since we're not using circular buffers anymore, so make sure to take it out

zsunberg commented on src/beliefs/k_previous_observations.jl:
```diff
@@ -0,0 +1,28 @@
+    return obs_stacked
+end
+
+function initialize_belief{O}(bu::KMarkovUpdater, obs_vec::Vector{O})
```
I don't think the `O` parameter is needed here, right?

MaximeBouton replied:
Should it just be `::Vector{Any}`?
Regardless the parameterization, there might be an ambiguity if O is a vector between:
```julia
initialize_belief{O}(bu::KMarkovUpdater, obs_vec::Vector{O})
initialize_belief{O}(bu::KMarkovUpdater, obs_vec::O)
```
I am not sure how to deal with that

zsunberg commented:
Try running it without doing it correctly and see what error message it puts out naturally (I'm guessing it will be a method error), then, if that error is not good enough, see where you can intercept it and make a better one.

I think maybe something like
```julia
update(bu::KMarkovUpdater, old_b, action, obs) = error("""
    KMarkovUpdater: Observation did not match previous observation type. Did you forget to pass the initial observation to the simulator?
   
    <example of how to do it>""")
```
I think that would catch the case where obs doesn't match the type of the other observations.

The other option is to be really permissive and have something like
```julia
new_container{O}(old_b::Vector{O}, obs::O) = Vector{O}
new_container(old_b::Vector, obs) = Vector{Any}
function update(bu::KMarkovUpdater, old_b::Vector, action, obs)
    obs_stacked = new_container(old_b, obs)(bu.k)
    ...
end
```
That way, if they don't pass in the first observation, the vector of previous observations will just contain the initial state distribution in a Vector{Any} and they can either deal with that or figure out that they can pass in the initial observation on their own.

Not sure which one I like better. Which one do you like?

MaximeBouton commented:
I added the error handling and change the  initialization to use a vector, the tests should pass :)

MaximeBouton commented:
Yup I forgot the readme, it is updated now. I am working on the error message.

zsunberg commented:
Did you commit the last changes? It doesn't look like it

MaximeBouton commented:
How does that look now?

zsunberg commented:
Ok, looks good to me. we can merge once the tests pass
